lc-hax/Scripts/Core/HaxCamera.cs

```diff
@@ -0,0 +1,162 @@
+using UnityEngine.Rendering.HighDefinition;
+
+internal class HaxCamera : MonoBehaviour {
+    internal static HaxCamera? Instance { get; private set; }
+    internal KeyboardMovement? KeyboardMovement { get; private set; }
+    internal MousePan? MousePan { get; private set; }
+    internal GameObject? HaxCameraContainer { get; private set; }
+    internal GameObject? HaxCamAudioContainer { get; private set; }
+    internal Camera? CustomCamera { get; private set; }
+    internal AudioListener? HaxCamAudioListener { get; private set; }
+
+    void Awake() {
+        Instance = this;
+        _ = this.GetCamera();
+    }
+
+    void OnEnable() {
+        GameListener.OnGameStart += this.DisableCamera;
+        GameListener.OnGameEnd += this.DisableCamera;
+    }
+
+    void OnDisable() {
+        GameListener.OnGameStart -= this.DisableCamera;
+        GameListener.OnGameEnd -= this.DisableCamera;
+    }
+
+    internal void SetActive(bool active) {
+        if (Helper.StartOfRound is not StartOfRound startOfRound) return;
+        if (Helper.LocalPlayer is not PlayerControllerB player) return;
+        if (player.activeAudioListener is not AudioListener playerlistener) return;
+        if (this.HaxCameraContainer is null) return;
+        if (this.HaxCamAudioListener is not AudioListener haxListener) return;
+        if (this.GetCamera() is not Camera camera) return;
+
+        this.HaxCameraContainer.SetActive(active);
+
+        if (!this.HaxCameraContainer.activeSelf) {
+            playerlistener.enabled = true;
+            startOfRound.audioListener = playerlistener;
+            startOfRound.activeCamera.enabled = true;
+        }
+        else {
+            this.CopyFromCamera(startOfRound.activeCamera.transform, camera, startOfRound.activeCamera);
+            startOfRound.activeCamera.enabled = false;
+            playerlistener.enabled = false;
+            startOfRound.audioListener = haxListener;
+        }
+    }
+
+    void LateUpdate() {
+        if (this.HaxCameraContainer is null) return;
+        if (this.KeyboardMovement is null) return;
+        if (this.HaxCameraContainer.activeSelf) {
+            this.KeyboardMovement.IsPaused = Helper.LocalPlayer is { isTypingChat: true };
+        }
+    }
+
+    internal void DisableCamera() {
+        if (PhantomMod.Instance is PhantomMod phantom) {
+            phantom.DisablePhantom();
+        }
+
+        this.HaxCameraContainer?.SetActive(false);
+    }
+
+    internal Camera? GetCamera() {
+        if (this.CustomCamera != null) return this.CustomCamera;
+
+        this.HaxCameraContainer ??= new GameObject("lc-hax Camera Parent");
+        Camera newCamera = this.HaxCameraContainer.AddComponent<Camera>();
+
+        this.HaxCamAudioContainer ??= new GameObject("lc-hax Audio Listener");
+        this.HaxCamAudioListener = this.HaxCamAudioContainer.GetComponent<AudioListener>();
+        this.HaxCamAudioListener ??= this.HaxCamAudioContainer.AddComponent<AudioListener>();
```

Comment:
I don't understand the logic here. If `HaxCameraContainer` already has a `Camera` component, then aren't we adding extras?

Reply:
Nvm, you end up setting `CustomCamera` at the end.